### ISSUE TITLE
TypeVar strangeness in type definitions

### ISSUE THREAD

**Issue by mauro3** [JuliaLang/julia#10642]
Two bugs which I think are related to how TypeVars are handled in the type-definitions.  Probably will be fixed in #8974?

On 0.4 and 0.3.6
```julia
immutable TT2{Ar}
    fld
    function TT2()
        @show T = Type{eltype(Ar)}
        @show typeof(T)
        @show (T, Int... ) # no error
        @show di = Dict(:a => (T, Int ))  # no error
        @show di = Dict(:a => T ) # no error
        @show Dict(:a => (T, Int... ) ) # error!!
        new(1)
    end
end

TT2{Array{Int,1}}() # this works
TT2{Array{Int}}()  # errors
```
the last statement errors with
```
julia> TT2{Array{Int}}()  # errors
T = Type{eltype(Ar)} = Type{Int64}
typeof(T) = DataType
(T,Int...) = (Type{Int64},Int64...)
di = Dict(:a => (T,Int)) = Dict(:a=>(Type{Int64},Int64))
di = Dict(:a => T) = Dict(:a=>Type{Int64})
ERROR: MethodError: `convert` has no method matching convert(::Type{TypeVar}, ::Type{Type{Int64}})
This may have arisen from a call to the constructor TypeVar(...),
since type constructors fall back to convert methods.
Closest candidates are:
  convert{T}(::Type{T}, ::T)
 in convert at base.jl:46
 in setindex! at dict.jl:587
 in call at dict.jl:337
 in call at none:94
```

Then, maybe related:
```julia
immutable TT{Ar}
    fld
    function TT() 
        @show T = Type{eltype(Ar)}
        @show ((T,Int64...), Ar)  # ((_<:Type{T},Int64...),Array{Int64,N})
        @show ((T,Int64...), 1)   # ((Type{Int64},Int64...),1)
        # Bonus, this errors:
        # @show Dict(:a => ((T,Int64...), 1) ) # note, replacing 1 with Ar does not error
        new(1)
    end
end
```
on 0.4 (Commit 01a3216*) I get:
```
julia> TT{Array{Int,1}}()
T = Type{eltype(Ar)} = Type{Int64}
((T,Int64...),Ar) = ((Type{Int64},Int64...),Array{Int64,1})
((T,Int64...),1) = ((Type{Int64},Int64...),1)
TT{Array{Int64,1}}(1)

julia> TT{Array{Int}}()
T = Type{eltype(Ar)} = Type{Int64}
((T,Int64...),Ar) = ((_<:Type{T},Int64...),Array{Int64,N})
((T,Int64...),1) = ((Type{Int64},Int64...),1)
TT{Array{Int64,N}}(1)
```
so for the second one a `_<:Type{T}` appears instead of `Type{Int64}`.  Note, this works in 0.3.6.

**Comment by JeffBezanson** [comment]
This seems to be fixed now, I believe by 3074b708433a170c1749bbb14806b27e3370c579.

**Comment by mauro3** [comment]
Cool, thanks!
